Update contributing doc for hugo specifics. (#1394)
platformsh/platformsh-docs: CONTRIBUTING.md

In order to maintain a consistent style and voice throughout our documentation please try to follow these standards and conventions when filing Pull Requests against our documentation.

## Hugo & Markdown

Hugo (which we use for documentation) uses Markdown as its file format.  

* All files should end in `.md`. They should be all-lower-case and match the title of the page, or in some cases an abbridged version of it.

> * All pages should start with a title, denoted with # (first level header).

Reviewer: This should be updated for the YAML frontmatter.

Author: I'm not sure I know where you're referring to here

Reviewer: "Pages should start with a title denoted with #".  That's GitBook-ese.  Hugo uses the YAML frontmatter instead.

Author: Ah, okay gotcha

- `title`: default dependent on theme.  The default is "Note", "Info", or "Warning" depending on the `theme`.

```
{{< note theme="warning" title="Danger, Will Robinson" >}}
Lorem ipsum!
{{< /note >}}
```

### Images

The standard Markdown syntax for images remains the same, but it is possible to pass values to it that modify the appearance of the image

```
<!-- Normal image syntax -->
![alt-title](/images/current_image.png)

<!-- Scaled image -->
![alt-title](/images/current_image.png "0.5")

<!-- Scaled image displayed inline, such as an icon in a sentence -->
![alt-title](/images/current_image.png "0.1-inline")
```

### Videos & asciinema

Displaying videos and asciinema recordings requires their own shortcodes, but they have identical parameters:

```
<!-- Video -->
{{< video src="videos/management-console/new-project-creation.mp4" >}}

<!-- Asciinema -->
{{< asciinema src="videos/asciinema/verify-cli-extended.cast" >}}
```

### Codetabs

You can display tabbed content that will apply syntax highlighting to it where defined.

```
{{< codetabs >}}

---
title=Elasticsearch
file=static/files/fetch/examples/php/elasticsearch
highlight=php
---

<--->

---
title=Memcached

> file=none

Reviewer: ```suggestion
```

highlight=python
---

from jwcrypto import jws, jwk

{{< /codetabs >}}
```

The first tab *Elasticsearch* will read from a local file and highlight it for PHP, whereas the second tab will apply Python highlighting to the content below `---`. Individual tabs are separated with the `<--->` line.

`markdownify` is `false` by default, although everything placed below `---` will be Markdown-rendered properly. This attribute really only needs to be `true` in cases where the `file` you are using for the tab's content is itself a `.md` file.

> All of these parameters need not be present for `codetabs` to function properly. You could very easily only include `title` and `highlight` for a code snippet. 

Reviewer: ```suggestion
Not all of these parameters need to be present for `codetabs` to function properly.  The only required properties are `title` and `highlight`. 
```

* Always use inline links.
* Do not hard-wrap prose text.  Set your text editor to soft wrapping.
* Internal links should be absolute (starting with `/`) and link to a file ending in `.md`, not `.html`. That makes them easier to find when needed to update.
* Each page must contain a front matter section, whether it is a single or list (`_index.md`) file. See the **Front matter** section below for more details

> ### Front matter

Reviewer: I think this can be brought up a header level, and thereby avoid the 4th level headers below.

Author: 👍

In general, this project uses two types of files: "single" and "list". List pages are made using `_index.md` files in every subdirectory, whereas single pages are generated from every other page without that name.

Additionally, single and list pages are built differently depending on their `layout`, which is defined in `layouts/_default/(single|list).html`. For our documentation, list pages by default generate buttons for each page in that subdirectory, whereas single pages are built from the contents of that file and include a table of contents.

#### Adding pages

If you wanted to add a new page for a runtime to the **Languages** section, you would add the following file:

```text
src/
   ...
   languages/
      ...
      new_language.md
```

`new_language.md` should contain at least the following in its front matter at the top of the page.

```
---
title: "New language"
weight: 4
---
```

> `title` sets the `<h1>` header for the page title, and weight defines in what order items in this subsection should be listed in the sidebar. Here the new page will appear fourth under the **Languages** section.

Reviewer: To clarify: Is it higher number before lower number, or lower before high?  Make that explicit, because different systems do it differently, just to piss you off.

Author: lower number = higher in the sidebar. I'll  include

The new page will have the link `/languages/new_language.html`

#### Adding subsections with pages

If you are adding a new *subsection* to **Languages** section, you would instead need to add:

```text
src/
   ...
   languages/
      ...
      new_language/
        _index.md
        specific_page.md
```

`_index.md` is a list page for the `new_language` subsection, and it will still be found at `/languages/new_language.html` like above. If you would like a new subsection's list page to be build to the list default, you can define the front matter the same as listed above for the single page.

Chances are, however, you don't want `_index.md` to be a list of that subdirectory's pages (especially if it's a new language), but rather that it's treated as a single page would: display content and contain a table of contents. You can tell Hugo to do this by including the `layout` attribute in its front matter.

```
---
title: "New language"
weight: 6
layout: single
---
```

`_index.md` will now use the layout for single pages (`layouts/_default/single.html`).

#### Sidebar and pages

> If you would like the title of a given page to appear differently in the sidebar - say, a shorter title there, but a more descriptive title on the page). You can use the `sidebarTitle` attribute in the front matter to define it, while the regular `title` attribute will only be used on the page itself.

Reviewer: ```suggestion
If you would like the title of a given page to appear differently in the sidebar &mdash; say, a shorter title there, but a more descriptive title on the page &mdash; you can use the `sidebarTitle` attribute in the front matter to define it, while the regular `title` attribute will only be used on the page itself.
```

```
---
title: "New language support details"    # Title on page
sidebarTitle: "New language"             # How title appears in sidebar
weight: 6
layout: single
---
```

#### Disabling table of contents

By default, all pages that use the `single` layout will generate a table of contents beneath their title. If you would like to disable it, add `toc` attribute to your front matter:

```
---
title: "New language"
weight: 4
toc: false
---
```

### Shortcodes

There are a few shortcodes written for the documentation that keep styling consistent and insert specific HTML into pages where they are referenced.

#### Notes

For the moment, Hugo does not provide a way to pass values to Markdown's blockquote syntax, but in this documentation there are a few variations on the blockquote that are used for different purposes. So, when leaving a "Note" or blockquote on your pages, use the `note` shortcode instead

```
{{< note >}}
Lorem ipsum
{{< /note }}
```

This shortcode will automatically create the **Note:** first line for you.

`note` accepts two optional parameters:

> - `theme`: default none, standard note. (`info`, `warning`). This will change the color and the default first line.

Reviewer: ```suggestion
- `theme`: May be one of `note, `info`, or `warning`.  Default is `note`.  This will change the color and the default first line.
```

> - `title`: default dependent on theme. Over-rides that default.

Reviewer: ```suggestion
- `title`: default dependent on theme.  The default is "Note", "Info", or "Warning" depending on the `theme`.
```

```
{{< note theme="warning" title="Danger, Will Robinson" >}}
Lorem ipsum!
{{< /note >}}
```

#### Images

The standard Markdown syntax for images remains the same, but it is possible to pass values to it that modify the appearance of the image

```
<!-- Normal image syntax -->
![alt-title](/images/current_image.png)

<!-- Scaled image -->
![alt-title](/images/current_image.png "0.5")

<!-- Scaled image displayed inline, such as an icon in a sentence -->
![alt-title](/images/current_image.png "0.1-inline")
```

#### Videos & asciinema

> Displaying videos and asciinema recordings requires their own shortcodes, but they have identical parameters:

Reviewer: Are there parameters beyond src?  (HTML video has a whole crapload of other options.)

Author: At the moment no. What would you  want to  add?

Reviewer: At minimum I think `type` is required, but there's a whole lot more: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video

Ideally we'd just allow these to pass-through rather than explicitly supporting each of them?  (Only applies to video, not asciinema, which has a different set of possible attributes.)

Author: Ya right now it's just set firm at `type="video/mp4"`. I'd say we set a few reasonable defaults and then accept everything to be overwritten.

```
<!-- Video -->
{{< video src="videos/management-console/new-project-creation.mp4" >}}

<!-- Asciinema -->
{{< asciinema src="videos/asciinema/verify-cli-extended.cast" >}}
```

#### Codetabs

You can display tabbed content that will apply syntax highlighting to it where defined.

```
{{< codetabs >}}

---
title=Elasticsearch
file=static/files/fetch/examples/php/elasticsearch
highlight=php
markdownify=false
---

<--->

---
title=Memcached
file=none
highlight=python
markdownify=false
---

from jwcrypto import jws, jwk

{{< /codetabs >}}
```

> The first tab *Elasticsearch* will read from a local file and highlight it for PHP, whereas the second tab will apply Python highlighting to the content below `---`. Individual tabs are separated with the `<--->` line.

Reviewer: What's markdownify all about, and what is the default?  (The default seems to be true, since everything I've seen specifies false, which feels backwards.)

Author: It will apply markdown by default to everything, except if the `file` you're using is a markdown file. Then you have to explicitly say make it markdown.
